packages-backend/express/src/types.ts

```diff
@@ -25,6 +25,17 @@ export type TSessionMiddlewareOptions = {
 
   /* Options for the `jwksRsa.expressJwtSecret` */
   jwks?: Omit<ExpressJwtOptions, 'jwksUri'>;
+
+  // In order to determine the 'audience' for JWT validation, the middleware needs
+  // to resolve the request url.
+  // By default, the middleware assumes it receives a Node.js IncomingMessage object
+  // as its first parameter.
+  // This default behaviour might not work for every scenario as, for example,
+  // AWS lambda functions, where that first parameter is an Event object.
+  // (Reference: https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html)
+  // To handle these situations, this option allows the delegation of the URL resolution
+  // from the request-like object provided to the middleware as its first parameter.
+  urlResolver?: (request: unknown) => string;
```

Comment:
I would keep it a bit simpler. Also I would maybe name the function something like `rewriteRequestUrl` (which we also use internally in the MC API).

```suggestion
  // By default we assume that the `request` is a Node.js-like object having either a `originalUrl` or `url` properties.
  // If that's not the case (for example in AWS Lambda functions) you need to correctly map the URL of
  // the `request` to return the full URI path, including query string parameters. 
  rewriteRequestUrl?: (request: unknown) => string;
```

On another note, instead of using `unknown` we can (re)use the generic type that we defined here:
https://github.com/commercetools/merchant-center-application-kit/blob/39a387cc507f244b23507a0efd80166e36038f68/packages-backend/express/src/auth.ts#L120

Then the `TSessionMiddlewareOptions` would also need to accept the generic type argument

```ts
export type TSessionMiddlewareOptions<Request extends IncomingMessage> {
  rewriteRequestUrl?: (request: Request) => string;
}
```

Reply:
I'm not sure about providing a type which needs to extend `IncomingMessage`. In the case of AWS lambda function, the type of the object received will have [this shape](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a1260a1f3d40f239b53fd29effba594b0d1bee08/types/aws-lambda/trigger/api-gateway-proxy.d.ts#L169) which does not extend [`IncomingMessage`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/http.d.ts#L785) so I guess consumers using TS for their source code will get a typing error.

The problem is that different providers could use totally different types so I'm unsure if we can provide any type here.

Reply:
> I'm not sure about providing a type which needs to extend `IncomingMessage`.

Ah yes, you're right. I'm wondering then if it would make sense to adjust that as well. So instead of requiring `Request extends IncomingMessage` we can create our own "base" request type with the minimal properties that we expect to be used.

```ts
interface TBaseRequest {
  headers: Record<string, undefined>,
  originalUrl?: string,
  url?: string,
}

interface TBaseResponse {}
```

> The way it is right now, this function is not expected to do that but return the URL.

True, you're right. How about `getRequestUrl`?

Reply:
I think that the nature of this optional function is to deal with contexts we don't know about. I mean, if the received first middleware parameter is what we expect, it will include know attributes (`url` or `originalUrl`). But if it's not, then we can't know what the shape of the object would be, unless we prepare our code to all different providers we support: in that case, we wouldn't need this option because our code would handle it.

That's why I think this function parameter shouldn't be typed. We want to support scenarios we don't know about. 
I mean, for know scenarios, our code will work. For unknown scenarios, we just don't have a clue about the received object so we totally delegate the URL resolution up to the consumers of the middleware.

Does it make sense?



> True, you're right. How about getRequestUrl?

Sure! I'll change it.

Reply:
Yes, you're right. I didn't noticed we're at least expecting the request-like object to have the `headers`.

I updated the types in this commit 51c8b2e915793a334a26f9d9c8b3db796e95fac7